src/ngamsServer/ngamsServer/ngamsFileUtils.py

```diff
@@ -72,6 +72,169 @@
+
+    Returns:
+
+    host:           Partner site host name (or IP address)
+
+    port:           Partner site port number
+
+    status_info:    Status response object (ngamsStatus)
+
+    disk_info:      Status response disk information object (ngamsDiskInfo)
+
+    file_info:      Status response file information object (ngamsFileInfo)
+    """
+    # If the request came from a partner site. We will not continue to
+    # propagate the request to avoid a death loop scenario. We will raise an
+    # exception.
+    if request_properties.hasHttpPar("partner_site_redirect"):
+        error_message = genLog("NGAMS_ER_UNAVAIL_FILE", [file_id])
+        logger.debug(error_message)
+        raise Exception(error_message)
+
+    # Check partner sites is enabled are available from the configuration
+    if not ngas_server.is_partner_sites_proxy_mode()\
+            or not ngas_server.get_partner_sites_address_list():
+        error_message = genLog("NGAMS_ER_UNAVAIL_FILE", [file_id])
+        logger.debug(error_message)
+        raise Exception(error_message)
+
+    # Lets query the partner sites for the availability of the requested file
+    authentication_header = ngamsSrvUtils.genIntAuthHdr(ngas_server)
+    parameter_list = [["file_id", file_id]]
+    if file_version != -1:
+        parameter_list.append(["file_version", file_version])
+    parameter_list.append(["partner_site_redirect", 1])
+
+    host, port, status_info, disk_info, file_info = None, None, None, None, None
+    for partner_site in ngas_server.get_partner_sites_address_list():
+        partner_address = partner_site.split(":")[0]
+        partner_port = int(partner_site.split(":")[-1])
+        try:
+            logger.info("Looking up file ID %s on partner site %s", file_id,
+                        partner_site)
+            response = ngamsHttpUtils.httpGet(partner_address, partner_port,
+                                              NGAMS_STATUS_CMD,
+                                              parameter_list,
+                                              auth=authentication_header)
+            with contextlib.closing(response):
+                response_info = response.read()
+        except:
+            # We ignore this error, and try the next partner site, if any
+            continue
+
+        status_info = ngamsStatus.ngamsStatus().unpackXmlDoc(response_info, 1)
+        logger.info("Result of File Access Query: {}".format(re.sub("\n", "",
+                        str(status_info.genXml().toprettyxml('  ', '\n')))))
+        if status_info.getStatus() == "FAILURE":
+            logger.info(genLog("NGAMS_INFO_FILE_NOT_AVAIL", [file_id, partner_address]))
+        else:
+            logger.info(genLog("NGAMS_INFO_FILE_AVAIL", [file_id, partner_address]))
+            disk_info = status_info.getDiskStatusList()[0]
+            file_info = disk_info.getFileObjList()[0]
+            host = partner_address
+            port = partner_port
+            break
```

Comment:
This of course looks a bit similar to what `ngamsFileUtils.locateArchiveFile` does internally, it would be interesting to see how much code can be shared here. I'm happy to do that in the future though.

Reply:
I think the fundamental difference is that `locateArchiveFile()` is query the NGAS DB to locate the file where as `lookup_partner_site_file()` is performing a HTTP query to the remote NGAS host. I think they populate the same status object in the end but using different sources.

Reply:
> @smclay thanks for the patches, it looks great :). The build is failing though, apparently because of a small change in one of the messages returned by the server. Could you have a look at that?
> 
> I also didn't see any new tests that exercise the new code, do you think you could add one? After we have tests passing in Travis we should get some measurements on code coverage, and could see what parts of the new code are getting hit and which aren't.
> 

I do not build NGAS. I assume this is done using the `build.sh`. I will test this and fix the problem.

You are correct. We need to write a test for this feature. Can you point me to a good example that I can use to get started? Is there a test that starts multiple NGAS servers where I can recreate a partner site setup?

Reply:
> I think the fundamental difference is that locateArchiveFile() is query the NGAS DB to locate the file where as lookup_partner_site_file() is performing a HTTP query to the remote NGAS host. I think they populate the same status object in the end but using different sources.

Yes, that's indeed the case. What I meant is that one of the things `locateArchiveFile` does internally is already very similar to what you are doing in `lookup_partner_site_file`: https://github.com/ICRAR/ngas/blob/c6556bcddf957751dc84556da5417ec4b9fd8fac/src/ngamsServer/ngamsServer/ngamsFileUtils.py#L258-L269

As I said though this is just me pointing out a potential for code reuse in the future, not something that needs to be tackled now.